File: tests/e2e/review_test.go
package e2e_test

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"gitlab.com/2finance/2finance-network/blockchain/contract/reviewV1"
	reviewV1Domain "gitlab.com/2finance/2finance-network/blockchain/contract/reviewV1/domain"
	"gitlab.com/2finance/2finance-network/blockchain/log"
	"gitlab.com/2finance/2finance-network/blockchain/utils"
	reviewV1Models "gitlab.com/2finance/2finance-network/blockchain/contract/reviewV1/models"
)

func TestReviewFlow(t *testing.T) {
	c := setupClient(t)
	reviewer, reviewerPriv := createWallet(t, c)
	reviewee, _ := createWallet(t, c)
	c.SetPrivateKey(reviewerPriv)

	deployedContract, err := c.DeployContract1(reviewV1.REVIEW_CONTRACT_V1)
	if err != nil {
		t.Fatalf("DeployContract: %v", err)
	}
	contractLog, err := utils.UnmarshalLog[log.Log](deployedContract.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (AddWallet.Logs[0]): %v", err)
	}

	// ------------------
	//   CREATE REVIEW
	// ------------------
	address := contractLog.ContractAddress
	subjectType := "order"
	subjectId := "order-xyz"
	rating := 5
	comment := "Great experience!"
	tags := map[string]string{"quality": "5"}
	mediaHashes := []string{"bafy1"}
	startAt := time.Now().Add(1 * time.Second)
	expiredAt := time.Now().Add(24 * time.Hour)
	hidden := false

	addReview, err := c.AddReview(address, reviewer.PublicKey, reviewee.PublicKey, subjectType, subjectId, rating, comment, tags, mediaHashes, startAt, expiredAt, hidden)
	if err != nil {
		t.Fatalf("AddReview: %v", err)
	}

	unmarshalLogReview, err := utils.UnmarshalLog[log.Log](addReview.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (AddReview.Logs[0]): %v", err)
	}
	assert.Equal(t, unmarshalLogReview.LogType, reviewV1Domain.REVIEW_ADDED_LOG)

	review, err := utils.UnmarshalEvent[reviewV1Domain.Review](unmarshalLogReview.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (AddReview.Logs[0].Event): %v", err)
	}

	assert.Equal(t, address, review.Address)
	assert.Equal(t, reviewer.PublicKey, review.Reviewer)
	assert.Equal(t, reviewee.PublicKey, review.Reviewee)
	assert.Equal(t, subjectType, review.SubjectType)
	assert.Equal(t, subjectId, review.SubjectID)
	assert.Equal(t, rating, review.Rating)
	assert.Equal(t, comment, review.Comment)
	assert.Equal(t, tags["quality"], review.Tags["quality"])
	assert.Equal(t, mediaHashes[0], review.MediaHashes[0])
	assert.WithinDuration(t, startAt, review.StartAt, time.Second)
	assert.WithinDuration(t, expiredAt, review.ExpiredAt, time.Second)
	assert.Equal(t, hidden, review.Hidden)

	getReview, err := c.GetReview(address)
	if err != nil {
		t.Fatalf("GetReview: %v", err)
	}

	var reviewState reviewV1Models.ReviewStateModel
	err = utils.UnmarshalState[reviewV1Models.ReviewStateModel](getReview.States[0].Object, &reviewState)
	if err != nil {
		t.Fatalf("UnmarshalState: %v", err)
	}

	assert.Equal(t, address, reviewState.Address)
	assert.Equal(t, reviewer.PublicKey, reviewState.Reviewer)
	assert.Equal(t, reviewee.PublicKey, reviewState.Reviewee)
	assert.Equal(t, subjectType, reviewState.SubjectType)
	assert.Equal(t, subjectId, reviewState.SubjectID)
	assert.Equal(t, rating, reviewState.Rating)
	assert.Equal(t, comment, reviewState.Comment)
	assert.Equal(t, tags["quality"], reviewState.Tags["quality"])
	assert.Equal(t, mediaHashes[0], reviewState.MediaHashes[0])
	assert.WithinDuration(t, startAt, reviewState.StartAt, time.Second)
	assert.WithinDuration(t, expiredAt, reviewState.ExpiredAt, time.Second)
	assert.Equal(t, hidden, reviewState.Hidden)

	// ------------------
	//   UPDATE REVIEW
	// ------------------
	newStart := time.Now()
	newExp := time.Now().Add(48 * time.Hour)

	updateReview, err := c.UpdateReview(address, subjectType, subjectId, 4, "Updated comment", map[string]string{"quality": "4"}, []string{"bafy2"}, &newStart, &newExp)
	if err != nil {
		t.Fatalf("UpdateReview: %v", err)
	}

	unmarshalLogUpdate, err := utils.UnmarshalLog[log.Log](updateReview.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (UpdateReview.Logs[0]): %v", err)
	}
	assert.Equal(t, unmarshalLogUpdate.LogType, reviewV1Domain.REVIEW_UPDATED_LOG)

	updatedReview, err := utils.UnmarshalEvent[reviewV1Domain.Review](unmarshalLogUpdate.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (UpdateReview.Logs[0].Event): %v", err)
	}

	assert.Equal(t, address, updatedReview.Address)
	assert.Equal(t, subjectType, updatedReview.SubjectType)
	assert.Equal(t, subjectId, updatedReview.SubjectID)
	assert.Equal(t, 4, updatedReview.Rating)
	assert.Equal(t, "Updated comment", updatedReview.Comment)
	assert.Equal(t, "4", updatedReview.Tags["quality"])
	assert.Equal(t, "bafy2", updatedReview.MediaHashes[0])

	getReview2, err := c.GetReview(address)
	if err != nil {
		t.Fatalf("GetReview (after update): %v", err)
	}

	var reviewState2 reviewV1Models.ReviewStateModel
	err = utils.UnmarshalState[reviewV1Models.ReviewStateModel](getReview2.States[0].Object, &reviewState2)
	if err != nil {
		t.Fatalf("UnmarshalState (after update): %v", err)
	}

	assert.Equal(t, address, reviewState2.Address)
	assert.Equal(t, reviewer.PublicKey, reviewState2.Reviewer)
	assert.Equal(t, reviewee.PublicKey, reviewState2.Reviewee)
	assert.Equal(t, subjectType, reviewState2.SubjectType)
	assert.Equal(t, subjectId, reviewState2.SubjectID)
	assert.Equal(t, 4, reviewState2.Rating)
	assert.Equal(t, "Updated comment", reviewState2.Comment)
	assert.Equal(t, "4", reviewState2.Tags["quality"])
	assert.Equal(t, "bafy2", reviewState2.MediaHashes[0])
	assert.WithinDuration(t, newStart, reviewState2.StartAt, time.Second)
	assert.WithinDuration(t, newExp, reviewState2.ExpiredAt, time.Second)

	// ------------------
	//   HIDE REVIEW
	// ------------------
	hideReview, err := c.HideReview(address, true)
	if err != nil {
		t.Fatalf("HideReview: %v", err)
	}

	unmarshalLogHide, err := utils.UnmarshalLog[log.Log](hideReview.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (HideReview.Logs[0]): %v", err)
	}
	assert.Equal(t, unmarshalLogHide.LogType, reviewV1Domain.REVIEW_HIDDEN_LOG)

	hiddenReview, err := utils.UnmarshalEvent[reviewV1Domain.Review](unmarshalLogHide.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (HideReview.Logs[0].Event): %v", err)
	}

	assert.Equal(t, address, hiddenReview.Address)
	assert.Equal(t, true, hiddenReview.Hidden)

	getReview3, err := c.GetReview(address)
	if err != nil {
		t.Fatalf("GetReview (after hide): %v", err)
	}

	var reviewState3 reviewV1Models.ReviewStateModel
	err = utils.UnmarshalState[reviewV1Models.ReviewStateModel](getReview3.States[0].Object, &reviewState3)
	if err != nil {
		t.Fatalf("UnmarshalState (after hide): %v", err)
	}

	assert.Equal(t, address, reviewState3.Address)
	assert.Equal(t, reviewer.PublicKey, reviewState3.Reviewer)
	assert.Equal(t, reviewee.PublicKey, reviewState3.Reviewee)
	assert.Equal(t, subjectType, reviewState3.SubjectType)
	assert.Equal(t, subjectId, reviewState3.SubjectID)
	assert.Equal(t, 4, reviewState3.Rating)
	assert.Equal(t, "Updated comment", reviewState3.Comment)
	assert.Equal(t, "4", reviewState3.Tags["quality"])
	assert.Equal(t, "bafy2", reviewState3.MediaHashes[0])
	assert.Equal(t, true, reviewState3.Hidden)

	// ------------------
	//   HELPFUL VOTE
	// ------------------
	voter, voterPriv := createWallet(t, c)
	c.SetPrivateKey(voterPriv)

	voteHelpful, err := c.VoteHelpful(address, voter.PublicKey, true)
	if err != nil {
		t.Fatalf("VoteHelpful: %v", err)
	}

	unmarshalLogVote, err := utils.UnmarshalLog[log.Log](voteHelpful.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (VoteHelpful.Logs[0]): %v", err)
	}
	assert.Equal(t, unmarshalLogVote.LogType, reviewV1Domain.REVIEW_HELPFUL_LOG)

	votedReview, err := utils.UnmarshalEvent[reviewV1Domain.Vote](unmarshalLogVote.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (VoteHelpful.Logs[0].Event): %v", err)
	}

	assert.Equal(t, address, votedReview.Address)
	assert.Equal(t, voter.PublicKey, votedReview.Voter)
	assert.Equal(t, true, votedReview.Helpful)

	getReview4, err := c.GetReview(address)
	if err != nil {
		t.Fatalf("GetReview (after helpful vote): %v", err)
	}

	var reviewState4 reviewV1Models.ReviewStateModel
	err = utils.UnmarshalState[reviewV1Models.ReviewStateModel](getReview4.States[0].Object, &reviewState4)
	if err != nil {
		t.Fatalf("UnmarshalState (after helpful vote): %v", err)
	}

	assert.Equal(t, address, reviewState4.Address)
	assert.Equal(t, reviewer.PublicKey, reviewState4.Reviewer)
	assert.Equal(t, reviewee.PublicKey, reviewState4.Reviewee)
	assert.Equal(t, subjectType, reviewState4.SubjectType)
	assert.Equal(t, subjectId, reviewState4.SubjectID)
	assert.Equal(t, 4, reviewState4.Rating)
	assert.Equal(t, "Updated comment", reviewState4.Comment)
	assert.Equal(t, "4", reviewState4.Tags["quality"])
	assert.Equal(t, "bafy2", reviewState4.MediaHashes[0])
	assert.Equal(t, true, reviewState4.Hidden)
	assert.Equal(t, true, reviewState4.HelpfulVotes[voter.PublicKey])

	// ------------------
	//   REPORT REVIEW
	// ------------------
	reporter, reporterPriv := createWallet(t, c)
	c.SetPrivateKey(reporterPriv)

	reportReview, err := c.ReportReview(address, reporter.PublicKey, "spam")
	if err != nil {
		t.Fatalf("ReportReview: %v", err)
	}

	unmarshalLogReport, err := utils.UnmarshalLog[log.Log](reportReview.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (ReportReview.Logs[0]): %v", err)
	}
	assert.Equal(t, unmarshalLogReport.LogType, reviewV1Domain.REVIEW_REPORTED_LOG)

	reportedReview, err := utils.UnmarshalEvent[reviewV1Domain.Report](unmarshalLogReport.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (ReportReview.Logs[0].Event): %v", err)
	}

	assert.Equal(t, address, reportedReview.Address)
	assert.Equal(t, reporter.PublicKey, reportedReview.Reporter)
	assert.Equal(t, "spam", reportedReview.Reason)

	getReview5, err := c.GetReview(address)
	if err != nil {
		t.Fatalf("GetReview (after report): %v", err)
	}

	var reviewState5 reviewV1Models.ReviewStateModel
	err = utils.UnmarshalState[reviewV1Models.ReviewStateModel](getReview5.States[0].Object, &reviewState5)
	if err != nil {
		t.Fatalf("UnmarshalState (after report): %v", err)
	}

	assert.Equal(t, address, reviewState5.Address)
	assert.Equal(t, reviewer.PublicKey, reviewState5.Reviewer)
	assert.Equal(t, reviewee.PublicKey, reviewState5.Reviewee)
	assert.Equal(t, subjectType, reviewState5.SubjectType)
	assert.Equal(t, subjectId, reviewState5.SubjectID)
	assert.Equal(t, 4, reviewState5.Rating)
	assert.Equal(t, "Updated comment", reviewState5.Comment)
	assert.Equal(t, "4", reviewState5.Tags["quality"])
	assert.Equal(t, "bafy2", reviewState5.MediaHashes[0])
	assert.Equal(t, true, reviewState5.Hidden)
	assert.NotEmpty(t, reviewState5.Reports)
	assert.Equal(t, reporter.PublicKey, reviewState5.Reports[0].Reporter)
	assert.Equal(t, "spam", reviewState5.Reports[0].Reason)

	// ------------------
	//   MODERATE REVIEW
	// ------------------
	c.SetPrivateKey(reviewerPriv)

	moderateReview, err := c.ModerateReview(address, reviewV1Domain.MODERATE_STATUS_APPROVED, "ok")
	if err != nil {
		t.Fatalf("ModerateReview: %v", err)
	}

	unmarshalLogModerate, err := utils.UnmarshalLog[log.Log](moderateReview.Logs[0])
	if err != nil {
		t.Fatalf("UnmarshalLog (ModerateReview.Logs[0]): %v", err)
	}
	assert.Equal(t, unmarshalLogModerate.LogType, reviewV1Domain.REVIEW_MODERATED_LOG)

	moderatedReview, err := utils.UnmarshalEvent[reviewV1Domain.Moderation](unmarshalLogModerate.Event)
	if err != nil {
		t.Fatalf("UnmarshalEvent (ModerateReview.Logs[0].Event): %v", err)
	}

	assert.Equal(t, address, moderatedReview.Address)
	assert.Equal(t, reviewV1Domain.MODERATE_STATUS_APPROVED, moderatedReview.Action)
	assert.Equal(t, "ok", moderatedReview.Note)

	getReview6, err := c.GetReview(address)
	if err != nil {
		t.Fatalf("GetReview (after moderate): %v", err)
	}

	var reviewState6 reviewV1Models.ReviewStateModel
	err = utils.UnmarshalState[reviewV1Models.ReviewStateModel](getReview6.States[0].Object, &reviewState6)
	if err != nil {
		t.Fatalf("UnmarshalState (after moderate): %v", err)
	}

	assert.Equal(t, address, reviewState6.Address)
	assert.Equal(t, reviewV1Domain.MODERATE_STATUS_APPROVED, reviewState6.ModerationStatus)
	assert.Equal(t, "ok", reviewState6.ModerationNote)
	assert.Equal(t, 4, reviewState6.Rating)
	assert.Equal(t, "Updated comment", reviewState6.Comment)
	assert.Equal(t, "4", reviewState6.Tags["quality"])
	assert.Equal(t, "bafy2", reviewState6.MediaHashes[0])
	assert.Equal(t, true, reviewState6.Hidden)

	// ------------------
	//   LIST REVIEWS
	// ------------------
	listReviews, err := c.ListReviews(reviewer.PublicKey, reviewee.PublicKey, subjectType, subjectId, nil, 0, 5, 1, 10, true)
	if err != nil {
		t.Fatalf("ListReviews: %v", err)
	}

	var reviews []reviewV1Domain.Review
	for _, state := range listReviews.States {
		var r reviewV1Models.ReviewStateModel
		err = utils.UnmarshalState[reviewV1Models.ReviewStateModel](state.Object, &r)
		if err != nil {
			t.Fatalf("UnmarshalState (ListReviews): %v", err)
		}
		reviews = append(reviews, reviewV1Domain.Review{
			Address:     r.Address,
			Reviewer:    r.Reviewer,
			Reviewee:    r.Reviewee,
			SubjectType: r.SubjectType,
			SubjectID:   r.SubjectID,
			Rating:      r.Rating,
			Comment:     r.Comment,
			Tags:        r.Tags,
			MediaHashes: r.MediaHashes,
			StartAt:     r.StartAt,
			ExpiredAt:   r.ExpiredAt,
			Hidden:      r.Hidden,
		})
	}

	assert.Len(t, reviews, 1)
	assert.Equal(t, address, reviews[0].Address)
	assert.Equal(t, reviewer.PublicKey, reviews[0].Reviewer)
	assert.Equal(t, reviewee.PublicKey, reviews[0].Reviewee)
	assert.Equal(t, subjectType, reviews[0].SubjectType)
	assert.Equal(t, subjectId, reviews[0].SubjectID)
	assert.Equal(t, 4, reviews[0].Rating)
	assert.Equal(t, "Updated comment", reviews[0].Comment)
	assert.Equal(t, "4", reviews[0].Tags["quality"])
	assert.Equal(t, "bafy2", reviews[0].MediaHashes[0])
	assert.WithinDuration(t, newStart, reviews[0].StartAt, time.Second)
	assert.WithinDuration(t, newExp, reviews[0].ExpiredAt, time.Second)
	assert.Equal(t, true, reviews[0].Hidden)
}

File: dart/blockchain/test/blockchain/contract/reviewV2/review_test.dart
import 'dart:convert';

import 'package:test/test.dart';

import 'package:two_finance_blockchain/blockchain/types/types.dart';
import 'package:two_finance_blockchain/blockchain/contract/reviewV2/constants.dart';
import 'package:two_finance_blockchain/blockchain/utils/marshal.dart';
import 'package:two_finance_blockchain/two_finance_blockchain.dart';
import '../../../helpers/helpers.dart' hide unmarshalEvent;

typedef JsonMap = Map<String, dynamic>;

Future<JsonMap> _getReviewState(TwoFinanceBlockchain c, String address) async {
  final out = await c.getReview(address: address);
  expect(out.states, isNotNull);
  expect(out.states!, isNotEmpty);

  return unmarshalState(
    out.states!.first.object,
    (json) => Map<String, dynamic>.from(json as Map),
  );
}

void _expectDateClose(
  String actualIso,
  DateTime expected, {
  int toleranceSeconds = 2,
}) {
  final actual = DateTime.parse(actualIso).toUtc();
  final expectedUtc = expected.toUtc();

  expect(
    actual.difference(expectedUtc).inSeconds.abs(),
    lessThanOrEqualTo(toleranceSeconds),
  );
}

void _expectReviewSnapshot(
  JsonMap review, {
  required String address,
  required String reviewer,
  required String reviewee,
  required String subjectType,
  required String subjectID,
  required int rating,
  required String comment,
  required String tagQuality,
  required String mediaHash,
  required bool hidden,
  required String moderationStatus,
  required String moderationNote,
  DateTime? startAt,
  DateTime? expiredAt,
}) {
  expect(review['address'], equals(address));
  expect(review['reviewer'], equals(reviewer));
  expect(review['reviewee'], equals(reviewee));
  expect(review['subject_type'], equals(subjectType));
  expect(review['subject_id'], equals(subjectID));
  expect(review['rating'], equals(rating));
  expect(review['comment'], equals(comment));
  expect(
    Map<String, dynamic>.from(review['tags'] as Map)['quality'],
    equals(tagQuality),
  );
  expect((review['media_hashes'] as List).first, equals(mediaHash));
  expect(review['hidden'], equals(hidden));
  expect(review['moderation_status'], equals(moderationStatus));
  expect(review['moderation_note'], equals(moderationNote));

  if (startAt != null) {
    _expectDateClose(review['start_at'] as String, startAt);
  }

  if (expiredAt != null) {
    _expectDateClose(review['expired_at'] as String, expiredAt);
  }
}

void main() {
  group('ReviewV2 E2E', () {
    test('E2E: deploy review contract + addReview + getReview', () async {
      final c = await setupClient();

      final reviewerKp = await validKeyPair();
      final reviewer = reviewerKp.publicKey;
      final revieweeKp = await validKeyPair();
      final reviewee = revieweeKp.publicKey;

      await c.setPrivateKey(reviewerKp.privateKey);

      final deployed = await c.deployContract1(REVIEW_CONTRACT_V2);
      expect(deployed.logs, isNotNull);
      expect(deployed.logs!, isNotEmpty);

      final reviewAddress = deployed.logs!.first.contractAddress;
      expect(reviewAddress, isNotEmpty);

      final subjectType = 'order';
      final subjectID = 'order-xyz';
      final rating = 5;
      final comment = 'Great experience!';
      final tags = <String, String>{'quality': '5'};
      final mediaHashes = <String>['bafy1'];
      final startAt = DateTime.now().toUtc().add(const Duration(seconds: 1));
      final expiredAt = DateTime.now().toUtc().add(const Duration(hours: 24));
      const hidden = false;

      final outAdd = await c.addReview(
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: rating,
        comment: comment,
        tags: tags,
        mediaHashes: mediaHashes,
        startAt: startAt,
        expiredAt: expiredAt,
        hidden: hidden,
      );

      expect(outAdd, isA<ContractOutput>());
      expect(outAdd.logs, isNotNull);
      expect(outAdd.logs!, isNotEmpty);

      final addLog = outAdd.logs!.first;
      expect(addLog.contractAddress, equals(reviewAddress));
      expect(addLog.logType, equals('Review_Added'));

      final addEvent = unmarshalEvent<JsonMap>(
        addLog.event,
        (json) => Map<String, dynamic>.from(json as Map),
      );

      expect(addEvent['address'], equals(reviewAddress));
      expect(addEvent['reviewer'], equals(reviewer));
      expect(addEvent['reviewee'], equals(reviewee));
      expect(addEvent['subject_type'], equals(subjectType));
      expect(addEvent['subject_id'], equals(subjectID));
      expect(addEvent['rating'], equals(rating));
      expect(addEvent['comment'], equals(comment));
      expect(
        Map<String, dynamic>.from(addEvent['tags'] as Map)['quality'],
        equals('5'),
      );
      expect((addEvent['media_hashes'] as List).first, equals('bafy1'));

      expect(addEvent.containsKey('hidden'), isFalse);
      expect(addEvent['moderation_status'], equals('pending'));

      _expectDateClose(addEvent['start_at'] as String, startAt);
      _expectDateClose(addEvent['expired_at'] as String, expiredAt);

      final reviewState = await _getReviewState(c, reviewAddress);

      _expectReviewSnapshot(
        reviewState,
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: rating,
        comment: comment,
        tagQuality: '5',
        mediaHash: 'bafy1',
        hidden: false,
        moderationStatus: 'pending',
        moderationNote: '',
        startAt: startAt,
        expiredAt: expiredAt,
      );

      expect(reviewState['helpful_votes'], isNull);
      expect(reviewState['reports'], isNull);

      // ------------------
      // UPDATE REVIEW
      // ------------------
      final newStart = DateTime.now().toUtc();
      final newExpiredAt = DateTime.now().toUtc().add(
        const Duration(hours: 48),
      );

      final outUpdate = await c.updateReview(
        address: reviewAddress,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: 4,
        comment: 'Updated comment',
        tags: const <String, String>{'quality': '4'},
        mediaHashes: const <String>['bafy2'],
        startAt: newStart,
        expiredAt: newExpiredAt,
      );

      expect(outUpdate, isA<ContractOutput>());
      expect(outUpdate.logs, isNotNull);
      expect(outUpdate.logs!, isNotEmpty);

      final updateLog = outUpdate.logs!.first;
      expect(updateLog.contractAddress, equals(reviewAddress));
      expect(updateLog.logType, equals('Review_Updated'));

      final updateEvent = unmarshalEvent<JsonMap>(
        updateLog.event,
        (json) => Map<String, dynamic>.from(json as Map),
      );

      expect(updateEvent['address'], equals(reviewAddress));
      expect(updateEvent['subject_type'], equals(subjectType));
      expect(updateEvent['subject_id'], equals(subjectID));
      expect(updateEvent['rating'], equals(4));
      expect(updateEvent['comment'], equals('Updated comment'));
      expect(
        Map<String, dynamic>.from(updateEvent['tags'] as Map)['quality'],
        equals('4'),
      );
      expect((updateEvent['media_hashes'] as List).first, equals('bafy2'));

      final reviewStateAfterUpdate = await _getReviewState(c, reviewAddress);

      _expectReviewSnapshot(
        reviewStateAfterUpdate,
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: 4,
        comment: 'Updated comment',
        tagQuality: '4',
        mediaHash: 'bafy2',
        hidden: false,
        moderationStatus: 'pending',
        moderationNote: '',
        startAt: newStart,
        expiredAt: newExpiredAt,
      );

      expect(reviewStateAfterUpdate['helpful_votes'], isNull);
      expect(reviewStateAfterUpdate['reports'], isNull);

      // ------------------
      // HIDE REVIEW
      // ------------------
      final outHide = await c.hideReview(reviewAddress, true);

      expect(outHide, isA<ContractOutput>());
      expect(outHide.logs, isNotNull);
      expect(outHide.logs!, isNotEmpty);

      final hideLog = outHide.logs!.first;
      expect(hideLog.contractAddress, equals(reviewAddress));
      expect(hideLog.logType, equals('Review_Hidden'));

      final hideEvent = unmarshalEvent<JsonMap>(
        hideLog.event,
        (json) => Map<String, dynamic>.from(json as Map),
      );

      expect(hideEvent['address'], equals(reviewAddress));
      expect(hideEvent['hidden'], isTrue);

      final reviewStateAfterHide = await _getReviewState(c, reviewAddress);

      _expectReviewSnapshot(
        reviewStateAfterHide,
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: 4,
        comment: 'Updated comment',
        tagQuality: '4',
        mediaHash: 'bafy2',
        hidden: true,
        moderationStatus: 'pending',
        moderationNote: '',
      );

      // ------------------
      // HELPFUL VOTE
      // ------------------
      final voterKp = await validKeyPair();
      final voter = voterKp.publicKey;

      await c.setPrivateKey(voterKp.privateKey);

      final outVoteHelpful = await c.voteHelpful(reviewAddress, voter, true);

      expect(outVoteHelpful, isA<ContractOutput>());
      expect(outVoteHelpful.logs, isNotNull);
      expect(outVoteHelpful.logs!, isNotEmpty);

      final voteLog = outVoteHelpful.logs!.first;
      expect(voteLog.contractAddress, equals(reviewAddress));
      expect(voteLog.logType, equals('Review_Helpful_Voted'));

      final voteEvent = unmarshalEvent<JsonMap>(
        voteLog.event,
        (json) => Map<String, dynamic>.from(json as Map),
      );

      expect(voteEvent['address'], equals(reviewAddress));
      expect(voteEvent['voter'], equals(voter));
      expect(voteEvent['helpful'], isTrue);

      final reviewStateAfterVote = await _getReviewState(c, reviewAddress);

      _expectReviewSnapshot(
        reviewStateAfterVote,
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: 4,
        comment: 'Updated comment',
        tagQuality: '4',
        mediaHash: 'bafy2',
        hidden: true,
        moderationStatus: 'pending',
        moderationNote: '',
      );

      final helpfulVotes = reviewStateAfterVote['helpful_votes'];
      expect(helpfulVotes, isNotNull);
      expect(Map<String, dynamic>.from(helpfulVotes as Map)[voter], isTrue);

      // ------------------
      // REPORT REVIEW
      // ------------------
      final reporterKp = await validKeyPair();
      final reporter = reporterKp.publicKey;

      await c.setPrivateKey(reporterKp.privateKey);

      final outReport = await c.reportReview(reviewAddress, reporter, 'spam');

      expect(outReport, isA<ContractOutput>());
      expect(outReport.logs, isNotNull);
      expect(outReport.logs!, isNotEmpty);

      final reportLog = outReport.logs!.first;
      expect(reportLog.contractAddress, equals(reviewAddress));
      expect(reportLog.logType, equals('Review_Reported'));

      final reportEvent = unmarshalEvent<JsonMap>(
        reportLog.event,
        (json) => Map<String, dynamic>.from(json as Map),
      );

      expect(reportEvent['address'], equals(reviewAddress));
      expect(reportEvent['reporter'], equals(reporter));
      expect(reportEvent['reason'], equals('spam'));

      final reviewStateAfterReport = await _getReviewState(c, reviewAddress);

      _expectReviewSnapshot(
        reviewStateAfterReport,
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: 4,
        comment: 'Updated comment',
        tagQuality: '4',
        mediaHash: 'bafy2',
        hidden: true,
        moderationStatus: 'pending',
        moderationNote: '',
      );

      final reports = reviewStateAfterReport['reports'];
      expect(reports, isNotNull);
      expect(reports, isA<List>());
      expect((reports as List), isNotEmpty);

      final firstReport = Map<String, dynamic>.from(reports.first as Map);
      expect(firstReport['reporter'], equals(reporter));
      expect(firstReport['reason'], equals('spam'));

      // ------------------
      // MODERATE REVIEW
      // ------------------
      await c.setPrivateKey(reviewerKp.privateKey);

      const approvedStatus = 'approved';

      final outModerate = await c.moderateReview(
        reviewAddress,
        approvedStatus,
        'ok',
      );

      expect(outModerate, isA<ContractOutput>());
      expect(outModerate.logs, isNotNull);
      expect(outModerate.logs!, isNotEmpty);

      final moderateLog = outModerate.logs!.first;
      expect(moderateLog.contractAddress, equals(reviewAddress));
      expect(moderateLog.logType, equals('Review_Moderated'));

      final moderateEvent = unmarshalEvent<JsonMap>(
        moderateLog.event,
        (json) => Map<String, dynamic>.from(json as Map),
      );

      expect(moderateEvent['address'], equals(reviewAddress));
      expect(moderateEvent['action'], equals(approvedStatus));
      expect(moderateEvent['note'], equals('ok'));

      final reviewStateAfterModerate = await _getReviewState(c, reviewAddress);

      _expectReviewSnapshot(
        reviewStateAfterModerate,
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: 4,
        comment: 'Updated comment',
        tagQuality: '4',
        mediaHash: 'bafy2',
        hidden: true,
        moderationStatus: approvedStatus,
        moderationNote: 'ok',
      );

      // ------------------
      // LIST REVIEWS
      // ------------------
      final outListReviews = await c.listReviews(
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        minRating: 0,
        maxRating: 5,
        page: 1,
        limit: 10,
        asc: true,
      );

      expect(outListReviews.states, isNotNull);
      expect(outListReviews.states!, isNotEmpty);

      final listedReviews = outListReviews.states!
          .map(
            (state) => unmarshalState(
              state.object,
              (json) => Map<String, dynamic>.from(json as Map),
            ),
          )
          .toList();

      expect(listedReviews.length, equals(1));

      final listedReview = listedReviews.first;

      _expectReviewSnapshot(
        listedReview,
        address: reviewAddress,
        reviewer: reviewer,
        reviewee: reviewee,
        subjectType: subjectType,
        subjectID: subjectID,
        rating: 4,
        comment: 'Updated comment',
        tagQuality: '4',
        mediaHash: 'bafy2',
        hidden: true,
        moderationStatus: approvedStatus,
        moderationNote: 'ok',
        startAt: newStart,
        expiredAt: newExpiredAt,
      );
    });
  });
}
